site/content/en/latest/design/security-policy.md

```diff
@@ -0,0 +1,114 @@
+
+## Non Goals
+* Define the API configuration fields in this API.
+
+## Implementation
+`SecurityPolicy` is a [Policy Attachment][] type API that can be used to extend [Gateway API][]
+to define authentication and authorization rules..
+
+### Example
+Here is an example highlighting how a user can configure this API.
+
+```
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: GatewayClass
+metadata:
+  name: eg
+spec:
+  controllerName: gateway.envoyproxy.io/gatewayclass-controller
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: eg
+  namespace: default
+spec:
+  gatewayClassName: eg
+  listeners:
+    - name: https
+      protocol: HTTPS
+      port: 443
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: backend
+  namespace: default
+spec:
+  parentRefs:
+    - name: eg
+  hostnames:
+    - "www.example.com"
+  rules:
+    - backendRefs:
+        - group: ""
+          kind: Service
+          name: backend
+          port: 3000
+          weight: 1
+      matches:
+        - path:
+            type: PathPrefix
+            value: /
+---
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: SecurityPolicy
+metadata:
+  name: jwt-authn-policy
+  namespace: default
+spec:
+  jwt:
+    providers:
+    - name: example
+      remoteJWKS:
+        uri: https://raw.githubusercontent.com/envoyproxy/gateway/main/examples/kubernetes/authn/jwks.json
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: eg
+    namespace: default
+```
+
+## Features / API Fields
+Here is a list of features that can be included in this API
+* JWT based authentication
+* OIDC Authentication
+* External Authorization
+* Basic Auth
+* API Key Auth
+* CORS
+
+## Design Decisions
+* This API will only support a single `targetRef` and can bind to a `Gateway` resource or a `HTTPRoute` or `GRPCRoute`.
+* This API resource MUST be part of same namespace as the targetRef resource
+* There can be only be ONE policy resource attached to a specific targetRef e.g. a `Listener` (section)  within a `Gateway`
+* If the policy targets a resource but cannot attach to it, this information should be reflected
+in the Policy Status field using the `Conflicted=True` condition.
+* If multiple polices target the same resource, the oldest resource (based on creation timestamp) will
+attach to the Gateway Listeners, the others will not.
+* If Policy A has a `targetRef` that includes a `sectionName` i.e. 
+it targets a specific Listener within a `Gateway` and Policy B has a `targetRef` that targets the same
+entire Gateway then
+  * Policy A will be applied/attached to the specific Listener defined in the `targetRef.SectionName`
+  * Policy B will be applied to the remaining Listeners within the Gateway. Policy B will have an additional
+  status condition `Overridden=True`.
+* A Policy targeting a xRoute (`HTTPRoute` or `GRPCRoute`) overrides a Policy targeting a Listener that is
```

Comment:
A policy targeting an xRoute may conflict with a policy targeting a Gateway. For example, the following two policies contains ext auth configurations with different grcp_service configured, but only one grpc_service can be set at the HCM filter level. 

Policy 1
```yaml
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: ext-auth-policy-gateway
  namespace: default
spec:
  extAuth:
    grpcService:
      envoyGrpc:
          clusterName: ext-authz-service1
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
    namespace: default
```

```yaml
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: ext-auth-policy-gateway
  namespace: default
spec:
  extAuth:
    grpcService:
      envoyGrpc:
          clusterName: ext-authz-service2
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: foo
    namespace: default   
```

We need to define how to handle this kind of conflict in the design docs. I can think of two rules:

* The oldest one wins.
* Or the one with the largest scope wins: Gateway > Listener > xRoute

Reply:
in the example you shared Policy 2 overrides Policy 1 and its entire spec wins since a xRoute target has high precedence over Gateway or Gateway section target (listener), so only Policy2's spec will be applied and translated

Reply:
There may be other HTTPRoutes. So Policy2 will be applied to HTTPRoute foo, and Policy1 should be applied to the owner listeners of foo as the default config for other HTTPRoutes. The conflict still exists.

A simpler example of a conflict:

HTTPRoute foo and bar target the same Gateway


```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: foo
  namespace: default
spec:
  parentRefs:
    - name: eg
  hostnames:
    - "www.example.com"
  rules:
    - backendRefs:
        - group: ""
          kind: Service
          name: backend-foo
          port: 3000
          weight: 1
      matches:
        - path:
            type: PathPrefix
            value: /foo
---

apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: bar
  namespace: default
spec:
  parentRefs:
    - name: eg
  hostnames:
    - "www.example.com"
  rules:
    - backendRefs:
        - group: ""
          kind: Service
          name: backend-bar
          port: 3000
          weight: 1
      matches:
        - path:
            type: PathPrefix
            value: /bar

```


Policy foo targets HTTPRoute foo, Policy bar targets HTTPRoute bar, and they have different extAuth configuration
```yaml
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: ext-auth-policy-foo
  namespace: default
spec:
  extAuth:
    grpcService:
      envoyGrpc:
          clusterName: ext-authz-service2
      ......
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: foo
    namespace: default 
---

apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: ext-auth-policy-bar
  namespace: default
spec:
  extAuth:
    grpcService:
      envoyGrpc:
          clusterName: ext-authz-service2
      ......
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: bar
    namespace: default   
```

In the Envoy configuration, only one grpc_service is allowed.

```yaml
http_filters:
  - name: envoy.filters.http.ext_authz
    typed_config:
      "@type": type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
      grpc_service:  // only one grpc service is allowed at the HCM filter level
        envoy_grpc:
          cluster_name: ext-authz-service1
      ......
```

JWT filter doesn't have this problem because the JWT filter configuration allows multiple providers configured.

```yaml
providers:
  provider1:
    issuer: https://provider1.com
    local_jwks:
      inline_string: PUBLIC-KEY
  provider2:
    issuer: https://provider2.com
    local_jwks:
      inline_string: PUBLIC-KEY
```

Reply:
can we create multiple ext authz filters using different names ?
https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/http_filters#route-specific-config